### PR TITLE
Adding Space Props

### PR DESCRIPTION
**Pull request by sharkparty**
added spacing props to top-level comps; and,
refactored naming convention

**Before submitting a pull request,** please make sure the following is done:

I have done **all** of the following:

- [x] Added a top level class to all my components `'.anchor-[COMPONENT NAME]'`.
- [x] Used [conventional commits](https://www.conventionalcommits.org) for all work.
- [x] Tested my solution on Mobile & Tablet.
- [x] Wrote [unit tests](https://jestjs.io/docs/en/getting-started) for states and all behavior (`npm test`) and passed coverage thresholds.
- [x] Updated snapshots for all permutations (`npm test -- -u`).
- [x] Accounted for hover, focus, blur, visited, & error states because they are not edge cases.
- [x] Created TODOs for known edge cases.
- [x] Documented all of my changes (inline & doc site).
- [x] Made sure that all accessibility errors are resolved.
- [x] Added [stories](https://storybook.js.org/docs/basics/introduction/) with knobs for all possible configurations.
- [x] De-linted and ran [prettier](https://github.com/prettier/prettier) (`npm run pretty`) on my code.
- [x] Added name to OWNERS file for all new components
- [x] If adding a new component, add its export to the rollup config

---------
**Outline your feature or bug-fix below**

My code removed padding from ListItems in a List; added spacing props to top-level comps; and, refactored naming convention

### CODE REVIEW

File: src/Typography/Typography.component.tsx [hunk 1]
```diff
@@ -139,13 +139,13 @@ const StyledTypography = (tag: TypographyTags) => styled(tag)<
 export const Typography = ({
     className,
     children,
-    tag = 'span',
+    tag: as = 'span',
     color,
     ...props
 }: TypographyProps): JSX.Element =>
-    React.createElement(StyledTypography(tag), {
+    React.createElement(StyledTypography(as), {
```
Comment (brandonlilly) [inline src/Typography/Typography.component.tsx]: Is `React.createElement` necessary anymore with using the `as` prop?

Also this line above
```tsx
const StyledTypography = (tag: TypographyTags) => styled(tag)<
```

File: src/Table/Table.component.tsx [hunk 2]
```diff
@@ -64,7 +66,8 @@ interface TableProps
         MarginProps,
```
Comment (brandonlilly) [inline src/Table/Table.component.tsx]: `MarginProps` are redundant now (same with `marginStyles` below)

File: src/Modal/Footer/__snapshots__/Footer.component.spec.tsx.snap [hunk 3]
```diff
@@ -23,6 +23,7 @@ exports[`Component: Modal.Footer should match snapshot with padding 1`] = `
   order: 1;
   justify-self: flex-end;
   background: transparent;
+  padding: 1rem;
```
Comment (brandonlilly) [inline src/Modal/Footer/__snapshots__/Footer.component.spec.tsx.snap:26]: Lol

File: src/Alert/__snapshots__/Alert.component.spec.tsx.snap [hunk 4]
```diff
@@ -48,7 +48,7 @@ Array [
   box-sizing: border-box;
   margin: 0;
   font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
-  font-size: 1rem;
+  font-size: 16px;
```
Comment (brandonlilly) [inline src/Alert/__snapshots__/Alert.component.spec.tsx.snap]: How did this happen?

File: src/Tabs/Tab/Tab.component.tsx [hunk 5]
```diff
@@ -2,11 +2,11 @@
 import * as React from 'react';
 import classNames from 'classnames';
 import styled, { css, ThemeContext } from '@xstyled/styled-components';
-import { th, space as spaceStyles } from '@xstyled/system';
+import { th, space as spaceStyles, SpaceProps } from '@xstyled/system';
 // ANCHOR
 import { Position, Variant, variantStyles } from '../utils';
 
-interface TabProps extends React.HTMLAttributes<HTMLDivElement> {
+interface TabProps extends React.HTMLAttributes<HTMLDivElement>, SpaceProps {
```
Comment (brandonlilly) [inline src/Tabs/Tab/Tab.component.tsx:9]: These are redundant now:
```tsx
    margin?: string;
    padding?: string;
```

File: src/Modal/Modal.component.tsx [hunk 6]
```diff
@@ -68,9 +68,10 @@ const StyledModal = StyledReactModal.default.styled`
         min-height: ${({ size = defaultSize }: ModalProps) =>
             Sizes[size].footerHeight}rem;
     }
+    ${spaceStyles}
```
Comment (brandonlilly) [inline src/Modal/Modal.component.tsx:69]: This ones a little more complicated with the 
```tsx
    margin: ${({ margin = 'auto' }: ModalProps) => margin};
```
above

File: src/Typography/Typography.component.tsx (hunk 1 above)
Comment (sharkparty) [reply src/Typography/Typography.component.tsx]: I was aiming for a partial refactor here, but it may make more sense for this to be a breaking change. Styled-Comps has supported a global `as` prop since V4 and we just don't need this anymore.

Comment (sharkparty) [reply src/Typography/Typography.component.tsx]: https://styled-components.com/docs/api#as-polymorphic-prop

File: src/Alert/__snapshots__/Alert.component.spec.tsx.snap (hunk 4 above)
Comment (sharkparty) [reply src/Alert/__snapshots__/Alert.component.spec.tsx.snap]: I'm not sure, but the updated one isn't in px.

File: src/Alert/__snapshots__/Alert.component.spec.tsx.snap [hunk 7]
```diff
@@ -2,7 +2,22 @@
 
 exports[`Component: Alert should match its snapshot. 1`] = `
 Array [
-  .c0 {
+  .c1 {
+  box-sizing: border-box;
+  margin: 0;
+  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
+  font-size: 1rem;
```
Comment (sharkparty) [inline src/Alert/__snapshots__/Alert.component.spec.tsx.snap:9]: Yay `rem`s! @brandonlilly

File: package.json [hunk 8]
```diff
@@ -41,9 +41,8 @@
     "build:modules": "rollup -c rollup.config.js",
     "ci": "concurrently --names 'lint,test' 'npm run lint' 'npm run test'",
     "prepublishOnly": "npm run build",
-    "lint": "concurrently --names 'pretty,tslint' 'npm run pretty' 'npm run tslint'",
```
Comment (brandonlilly) [inline package.json:44]: No more lint?

File: src/theme/typography.theme.ts [hunk 9]
```diff
@@ -115,6 +115,7 @@ export const typography: TypographyTheme = {
         h1: {
             fontSize: '1.75rem',
             lineHeight: '2rem',
+            color: 'red',
```
Comment (brandonlilly) [inline src/theme/typography.theme.ts]: ^

File: src/Typography/Typography.component.tsx [hunk 10]
```diff
@@ -139,14 +137,14 @@ const StyledTypography = (tag: TypographyTags) => styled(tag)<
 export const Typography = ({
     className,
     children,
-    tag = 'span',
+    as = 'span',
```
Comment (brandonlilly) [inline src/Typography/Typography.component.tsx]: I think specifying the default here should be redundant, since its a `span` normally. The only difference afaict would be that the prop isn't defined by default for the variant styles- but we can specify a default there anyway (which is currently set to `none`).

File: src/Typography/Typography.component.tsx [hunk 11]
```diff
@@ -139,14 +137,14 @@ const StyledTypography = (tag: TypographyTags) => styled(tag)<
 export const Typography = ({
     className,
     children,
-    tag = 'span',
```
Comment (brandonlilly) [inline src/Typography/Typography.component.tsx:142]: If we want to deprecate over time, we could accept `tag` pass it in to `as` and console.warn that `tag` is being deprecated.

File: src/Typography/Typography.component.tsx [hunk 12]
```diff
@@ -139,14 +137,14 @@ const StyledTypography = (tag: TypographyTags) => styled(tag)<
...
-        children,
-        tag,
-        $color: color,
-        className: classNames('anchor-typography', className),
-        ...props,
-    });
+}: TypographyProps): JSX.Element => (
+    <StyledTypography
+        className={classNames('anchor-typography', className)}
+        {...{ as, color, ...props }}
```
Comment (brandonlilly) [inline src/Typography/Typography.component.tsx]: `$color: color` keeps `color="text.base"` out of the dom.

you should be able to keep it out of this object spread and just do `$color={color}`

File: src/List/Title/__snapshots__/Title.component.spec.tsx.snap [hunk 13]
```diff
@@ -24,19 +10,34 @@ exports[`Component: Title should be defined 1`] = `
   font-weight: 500;
   text-align: inherit;
   color: inherit;
-  color: #808080;
```
Comment (brandonlilly) [inline src/List/Title/__snapshots__/Title.component.spec.tsx.snap]: This shouldn't be gone (see `$color` vs `color` comment)

File: src/List/Title/__snapshots__/Title.component.spec.tsx.snap [hunk 14]
```diff
@@ -24,19 +10,34 @@ exports[`Component: Title should be defined 1`] = `
...
+  height: 2rem;
+  margin-top: 0.75rem;
+}
+
 <div
   className="c0 anchor-list-title"
 >
   <span
     className="c1 anchor-typography"
+    color="text.label"
```
Comment (brandonlilly) [inline src/List/Title/__snapshots__/Title.component.spec.tsx.snap]: But this should be gone (see `$color` vs `color` comment)

File: src/AutoComplete/ResultsContainer/__snapshots__/ResultsContainer.component.spec.tsx.snap [hunk 15]
```diff
@@ -363,7 +290,8 @@ exports[`Component: ResultsContainer should be defined 1`] = `
       onMouseOver={[Function]}
     >
       <span
-        className="c5 anchor-typography c4 lg"
```
Comment (brandonlilly) [inline src/AutoComplete/ResultsContainer/__snapshots__/ResultsContainer.component.spec.tsx.snap:366]: I'm confused by why `anchor-typography` is disappearing.

File: package.json (hunk 8 above)
Comment (sharkparty) [reply package.json:44]: So tslint as a project is EOL'd, but lemme add eslint with the ts plugins.

File: src/theme/typography.theme.ts (hunk 9 above)
Comment (sharkparty) [reply src/theme/typography.theme.ts]: Nice catch.

File: src/Typography/Typography.component.tsx (hunk 11 above)
Comment (sharkparty) [reply src/Typography/Typography.component.tsx:142]: I was initially doing just that, but this has been included in the SC lib since v4 and we really just need to move this forward.

File: src/Typography/Typography.component.tsx (hunk 12 above)
Comment (sharkparty) [reply src/Typography/Typography.component.tsx]: Another artifact from my naive first SC.

File: src/List/Title/__snapshots__/Title.component.spec.tsx.snap (hunk 14 above)
Comment (sharkparty) [reply src/List/Title/__snapshots__/Title.component.spec.tsx.snap]: It is now, thanks!

File: package.json [hunk 16]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@retailmenot/anchor",
-  "version": "1.3.7",
+  "version": "1.4.0",
```
Comment (jcristol) [inline package.json:3]: just curious why jump a new minor version and a couple of bug fix versions?

File: src/utils/position/position.ts [hunk 17]
```diff
@@ -20,7 +20,7 @@ export const positionVariants = (
     width: number,
     overlayHeight: number,
     overlayWidth: number,
-    spacing: number = 0
+    spacing = 0
```
Comment (jcristol) [inline src/utils/position/position.ts:23]: was this removed for redundancy sake?

File: src/Tabs/Tabs.component.tsx [hunk 18]
```diff
@@ -94,12 +94,12 @@ export const Tabs = ({
     onChange,
     ...props
 }: TabsProps): React.ReactElement<TabsProps> => {
-    let tabInfo: Array<{
```
Comment (jcristol) [inline src/Tabs/Tabs.component.tsx:97]: why move from `Array` of blah to this other notation?

**Review by jcristol — COMMENTED** [review]
Added some review items mostly for my education. Skimmed through the Components because they all look like they similar change sets. But as a whole this makes sense.

File: package.json (hunk 16 above)
Comment (sharkparty) [reply package.json:3]: This introduces breaking changes with the Typog..y component, but it's merely a consistency fix with a lower level dependency so I felt like a good middle-ground would be a minor release.

File: src/utils/position/position.ts (hunk 17 above)
Comment (sharkparty) [reply src/utils/position/position.ts:23]: This is a lint rule I enabled. The tl;dr: is do not allow people to specify a type redundantly when it can be inferred from the instantiation.

File: src/Tabs/Tabs.component.tsx (hunk 18 above)
Comment (sharkparty) [reply src/Tabs/Tabs.component.tsx:97]: Hah, I actually kinda hate this. It's another lint rule that I enabled. While I hate this, the Typescript community is more widely swayed by the `[]int` syntax than `Array<int>` and the most compelling reason is that `Array` is a JS language construct that is being used by TS as an interface which is "confusing". I'm cool switching this back because I think it reads better, but this seems to be more widely adopted by the TS community, especially those in project we borrow capital from.

**Review by jcristol — APPROVED** [review]

**Review by jcristol — APPROVED** [review]

**Review by jcristol — APPROVED** [review]

**Review by jcristol — APPROVED** [review]

File: src/AutoComplete/ResultsContainer/__snapshots__/ResultsContainer.component.spec.tsx.snap [hunk 19]
```diff
@@ -156,7 +137,7 @@ exports[`Component: ResultsContainer should accept a resultTemplate 1`] = `
       onMouseOver={[Function]}
     >
       <span
-        className="c2 anchor-typography c3 lg"
+        className="sc-LzLte c2 lg anchor-typography"
```
Comment (brandonlilly) [inline src/AutoComplete/ResultsContainer/__snapshots__/ResultsContainer.component.spec.tsx.snap:140]: The change I made to the Typography components isn't playing 100% nice with the `jest-styled-components` render. Afaict, all of where this occurs has the same hash (`sc-LzLte`). The only real potential issue in my mind is test fragility (i.e. a simple change modifies the hash, requiring a bunch of snapshot update).

File: src/Typography/Typography.component.tsx [hunk 20]
```diff
@@ -88,9 +86,14 @@ export interface StyledTypographyProps extends TypographyProps {
     $color?: 'inherit' | 'initial' | string;
 }
 
-const StyledTypography = (tag: TypographyTags) => styled(tag)<
-    StyledTypographyProps
->`
+export const Typography = styled('span').attrs<TypographyProps>(
```
Comment (brandonlilly) [inline src/Typography/Typography.component.tsx:89]: Converting Typography to a `styled` component (rather than a component that returns a styled component) allows styled-components to squash the styles of components that styled-extend Typography (i.e.  styled(Typography)).

The issue with the missing `anchor-typography` classes was with components that both extend Typography (e.g. Hero.Title, Alert, List.Item) and use the `as` prop. The explanation is complicated (https://github.com/styled-components/styled-components/issues/1981#issuecomment-419287003) and is also the expected behavior for styled-components. This change fixes that issue.

Comment (sharkparty) [reply src/Typography/Typography.component.tsx:89]: I read through this and it's a known issue and while I don't like that we lose the ability to access typography by it's class under certain conditions, I don't think this is a gate for this PR.

Comment (sharkparty) [reply src/Typography/Typography.component.tsx:89]: Also, thanks for finding this and spending the time to explain it.

**Review by brandonlilly — APPROVED** [review]
LFTM
